### PR TITLE
Generalize the RHE Function beyond AIA

### PR DESCRIPTION
**Pull request by GillySpace27**

### CODE REVIEW

**Comment by GillySpace27** [comment]
This PR is associated with Issue #233 addressing the ability for the algorithm to operate on non-AIA datasets.

File: sunkit_image/radial.py [hunk 1]
```diff
@@ -221,7 +225,6 @@ def nrgf(
     width_function=np.std,
     width_function_kwargs=None,
     application_radius=1 * u.R_sun,
-    progress=True,
```
Comment (nabobalis) [inline sunkit_image/radial.py:224]: This changes will need to be undone.

File: sunkit_image/radial.py [hunk 2]
```diff
@@ -207,6 +209,8 @@ def intensity_enhance(
     enhancement[map_r < normalization_radius] = 1
 
     # Return a map with the intensity enhanced above the normalization radius
+    # and the same meta data as the input map.
+
```
Comment (nabobalis) [inline sunkit_image/radial.py:213]:
```suggestion
```

File: sunkit_image/radial.py [hunk 3]
```diff
@@ -351,7 +351,7 @@ def set_attenuation_coefficients(order, range_mean=None, range_std=None, cutoff=
 
     .. note::
 
-        The returned maps have their ``plot_settings`` changed to remove the extra normalization step.
+    The returned maps have their ``plot_settings`` changed to remove the extra normalization step.
```
Comment (nabobalis) [inline sunkit_image/radial.py:354]:
```suggestion
        The returned maps have their ``plot_settings`` changed to remove the extra normalization step.
```

File: sunkit_image/radial.py [hunk 4]
```diff
@@ -634,13 +630,14 @@ def _percentile_ranks_numpy_inplace(arr):
...
+    *,
     radial_bin_edges=None,
     application_radius=0 * u.R_sun,
     upsilon=0.35,
     method="numpy",
-    *,
-    vignette=1.5 * u.R_sun,
+    vignette=None,  # 1.5 * u.R_sun,
     progress=True,
+    animate=False,
```
Comment (nabobalis) [inline sunkit_image/radial.py:640]: The animate will need to go.

File: sunkit_image/radial.py [hunk 5]
```diff
@@ -634,13 +630,14 @@ def _percentile_ranks_numpy_inplace(arr):
...
 def rhef(
     smap,
+    *,
     radial_bin_edges=None,
     application_radius=0 * u.R_sun,
     upsilon=0.35,
     method="numpy",
-    *,
-    vignette=1.5 * u.R_sun,
+    vignette=None,  # 1.5 * u.R_sun,
```
Comment (nabobalis) [inline sunkit_image/radial.py:638]: These changes also need to be reverted or did you want to set this to be None?

File: sunkit_image/radial.py [hunk 6]
```diff
@@ -695,46 +687,69 @@ def rhef(
...
 
+        # If animation is enabled, update the image data in the figure
+        if animate:
+            im.set_data(data)
+            plt.pause(0.0001)  # Pause to refresh the plot
+    if animate:
+        plt.show(block=True)
+
+    # Final update to the map
+    new_map = Map(data, smap.meta, autoalign=True)
```
Comment (nabobalis) [inline sunkit_image/radial.py:744]: This autoalign does something?

File: sunkit_image/radial.py [hunk 7]
```diff
@@ -695,46 +687,69 @@ def rhef(
...
-        new_map = blackout_pixels_above_radius(new_map, vignette)
+        new_map = blackout_pixels_above_radius(new_map, vignette.to(u.R_sun))
 
+    # Adjust plot settings to remove extra normalization
     # This must be done whenever one is adjusting the overall statistical distribution of values
+
     new_map.plot_settings["norm"] = None
 
+    # Return the new SunPy map with RHEF applied
     return new_map
```
Comment (nabobalis) [inline sunkit_image/radial.py:755]:
```suggestion
    # Adjust plot settings to remove extra normalization
    # due to the difference in  the overall statistical distribution
    new_map.plot_settings["norm"] = None
    return new_map
```

File: sunkit_image/radial.py [hunk 8]
```diff
@@ -695,46 +687,69 @@ def rhef(
...
+        # If animation is enabled, update the image data in the figure
+        if animate:
+            im.set_data(data)
+            plt.pause(0.0001)  # Pause to refresh the plot
+    if animate:
+        plt.show(block=True)
+
+    # Final update to the map
+    new_map = Map(data, smap.meta, autoalign=True)
+    # Apply the vignette if specified
```
Comment (nabobalis) [inline sunkit_image/radial.py:745]:
```suggestion
```

**Comment by nabobalis** [comment]
Thanks for the PR @GillySpace27, what other data have were you trying to run this on? 

We will need to get some test data to confirm it works for non-AIA data.

File: sunkit_image/radial.py (hunk 5 above)
Comment (GillySpace27) [reply sunkit_image/radial.py:638]: This turned out to be one of the bugs I found: when working with data that is much wider FOV than AIA, the vignette should be none or set to the user's input. So I think defaulting to None is best, unless we wanted some instrument-specific/ metadata-specific value to be set.

File: sunkit_image/radial.py (hunk 4 above)
Comment (GillySpace27) [reply sunkit_image/radial.py:640]: That's fair, thats just how I found the error.

**Comment by GillySpace27** [comment]
@nabobalis Thanks! The first non-AIA dataset I applied it to is some synthetic PUNCH test data. I'm interested in getting some others working too, including LASCO, SUVI, etc.

File: sunkit_image/radial.py (hunk 6 above)
Comment (GillySpace27) [reply sunkit_image/radial.py:744]: Probably not, I just saw it in an example.

File: sunkit_image/radial.py (hunk 5 above)
Comment (nabobalis) [reply sunkit_image/radial.py:638]: Ah in that case yeah, set this to None. Only is a user sets the value, should we use it.

File: sunkit_image/radial.py [hunk 9]
```diff
@@ -634,12 +630,12 @@ def _percentile_ranks_numpy_inplace(arr):
...
 def rhef(
     smap,
+    *,
     radial_bin_edges=None,
     application_radius=0 * u.R_sun,
     upsilon=0.35,
     method="numpy",
-    *,
-    vignette=1.5 * u.R_sun,
+    vignette=None,  # 1.5 * u.R_sun,
```
Comment (nabobalis) [inline sunkit_image/radial.py:638]:
```suggestion
    vignette=None,
```

File: sunkit_image/utils/utils.py [hunk 10]
```diff
@@ -392,7 +394,7 @@ def blackout_pixels_above_radius(smap, radius_limit=1.5 * u.R_sun):
     mask = map_r > radius_limit
 
     # Apply the mask to the map data
-    masked_data = np.where(mask, 0, smap.data)
+    masked_data = np.where(mask, np.NAN, smap.data)
```
Comment (nabobalis) [inline sunkit_image/utils/utils.py:397]:
```suggestion
    masked_data = np.where(mask, np.nan, smap.data)
```
?

File: sunkit_image/radial.py [hunk 11]
```diff
@@ -695,46 +684,57 @@ def rhef(
...
+        new_map = blackout_pixels_above_radius(new_map, vignette.to(u.R_sun))
+
+    # Adjust plot settings to remove extra normalization
+    # This must be done whenever one is adjusting
+    # the overall statistical distribution of values
 
-    # This must be done whenever one is adjusting the overall statistical distribution of values
     new_map.plot_settings["norm"] = None
 
+    # Return the new SunPy map with RHEF applied
```
Comment (nabobalis) [inline sunkit_image/radial.py:739]:
```suggestion
```

File: sunkit_image/radial.py [hunk 12]
```diff
@@ -695,46 +684,57 @@ def rhef(
...
+    # Final update to the map
+    new_map = Map(data, smap.meta)
 
     if vignette is not None:
-        new_map = blackout_pixels_above_radius(new_map, vignette)
+        new_map = blackout_pixels_above_radius(new_map, vignette.to(u.R_sun))
+
+    # Adjust plot settings to remove extra normalization
+    # This must be done whenever one is adjusting
+    # the overall statistical distribution of values
```
Comment (nabobalis) [inline sunkit_image/radial.py:736]:
```suggestion
```

File: sunkit_image/radial.py [hunk 13]
```diff
@@ -695,46 +684,57 @@ def rhef(
...
-
-        # Perform the filtering operation
-        ranking_func = _select_rank_method(method)
+        # Apply ranking function
         data[here] = ranking_func(smap.data[here])
         if upsilon is not None:
             data[here] = apply_upsilon(data[here], upsilon)
-    new_map = sunpy.map.Map(data, meta, autoalign=True)
+
+    # Final update to the map
```
Comment (nabobalis) [inline sunkit_image/radial.py:727]:
```suggestion
```

File: sunkit_image/radial.py [hunk 14]
```diff
@@ -695,46 +684,57 @@ def rhef(
...
+        # Identify pixels within the current radial bin
+        here = np.logical_and(
+            map_r >= radial_bin_edges[0, i].to(u.R_sun), map_r < radial_bin_edges[1, i].to(u.R_sun)
+        )
         if application_radius is not None and application_radius > 0:
             here = np.logical_and(here, map_r >= application_radius)
-
-        # Perform the filtering operation
-        ranking_func = _select_rank_method(method)
+        # Apply ranking function
```
Comment (nabobalis) [inline sunkit_image/radial.py:722]:
```suggestion
```

File: sunkit_image/radial.py [hunk 15]
```diff
@@ -695,46 +684,57 @@ def rhef(
...
     # Allocate storage for the filtered data
     data = np.zeros_like(smap.data)
-    meta = smap.meta
 
+    # Enable progress bar for large numbers of bins
     if radial_bin_edges.shape[1] > 2000:
         progress = True
 
-    # Calculate the filter values for each radial bin.
+    # Select the ranking method
```
Comment (nabobalis) [inline sunkit_image/radial.py:711]:
```suggestion
```

File: sunkit_image/radial.py [hunk 16]
```diff
@@ -695,46 +684,57 @@ def rhef(
...
+            * u.R_sun
         )
-
     # Allocate storage for the filtered data
     data = np.zeros_like(smap.data)
-    meta = smap.meta
 
+    # Enable progress bar for large numbers of bins
     if radial_bin_edges.shape[1] > 2000:
         progress = True
```
Comment (nabobalis) [inline sunkit_image/radial.py:709]:
```suggestion
```
Should leave this up to the user only.

File: sunkit_image/radial.py [hunk 17]
```diff
@@ -695,46 +684,57 @@ def rhef(
...
+        radial_bin_edges = (
+            equally_spaced_bins(
+                inner_value=radial_bin_edges[0, 0].to(u.R_sun).value,
+                outer_value=np.max(map_r.to(u.R_sun)).value,
+                nbins=radial_bin_edges.shape[1],
+            )
+            * u.R_sun
         )
-
     # Allocate storage for the filtered data
```
Comment (nabobalis) [inline sunkit_image/radial.py:704]:
```suggestion
```

File: sunkit_image/radial.py [hunk 18]
```diff
@@ -695,46 +684,57 @@ def rhef(
       https://www.proquest.com/docview/2759080511
     """
 
-    # Get the radii for every pixel
-    map_r = find_pixel_radii(smap).to(u.R_sun)
+    # Get the radii for every pixel, ensuring units are correct (in terms of pixels or solar radii)
+    map_r = find_pixel_radii(smap)
 
+    # Automatically generate radial bin edges if none are provided
```
Comment (nabobalis) [inline sunkit_image/radial.py:690]:
```suggestion
```

File: sunkit_image/radial.py [hunk 19]
```diff
@@ -695,46 +684,57 @@ def rhef(
       https://www.proquest.com/docview/2759080511
     """
 
-    # Get the radii for every pixel
-    map_r = find_pixel_radii(smap).to(u.R_sun)
+    # Get the radii for every pixel, ensuring units are correct (in terms of pixels or solar radii)
```
Comment (nabobalis) [inline sunkit_image/radial.py:687]:
```suggestion
```

File: sunkit_image/radial.py [hunk 20]
```diff
@@ -658,33 +654,26 @@ def rhef(
...
-        If you want to disable this, pass in None.
-        Set pixels above this radius to black.
-        Defaults to None which is no vignette.
-        One suggested value is ``1.5*u.R_sun``.
-    progress: bool, optional
-        Display a progressbar on the main loop.
-        Defaults to True.
+    upsilon : float or None, optional
+        A double-sided gamma function to apply to modify the equalized histograms. Defaults to 0.35.
+    method : str, optional
```
Comment (nabobalis) [inline sunkit_image/radial.py:666]: Do we have a list of all the known methods?

File: sunkit_image/radial.py [hunk 21]
```diff
@@ -658,33 +654,26 @@ def rhef(
...
+        Method used to rank the pixels for equalization. Defaults to 'inplace'.
+    vignette : `astropy.units.Quantity`, optional
+        Radius beyond which pixels will be set to black. Defaults to ``1.5*u.R_sun``. Set to None to disable vignetting.
```
Comment (nabobalis) [inline sunkit_image/radial.py:669]:
```suggestion
        Radius beyond which pixels will be set to black. Defaults to None, must be in units that are compatible with "R_sun" as the value will be transformed. 
```

**Comment by nabobalis** [comment]
> @nabobalis Thanks! The first non-AIA dataset I applied it to is some synthetic PUNCH test data. I'm interested in getting some others working too, including LASCO, SUVI, etc.

Is it possible we can add a file from the synthetic PUNCH test data to the repo?

File: sunkit_image/radial.py (hunk 20 above)
Comment (GillySpace27) [reply sunkit_image/radial.py:666]: inplace, numpy, and scipy. They are defined above in a helper function. This functionality could probably be removed, but it is nice to have the flexibility I think.

File: sunkit_image/radial.py (hunk 21 above)
Comment (GillySpace27) [reply sunkit_image/radial.py:669]: I'm thinking that setting to NAN is probably more useful since the values aren't valid out there anyway.

Comment (nabobalis) [reply sunkit_image/radial.py:669]: Sure but you won't pass nan to the keyword?

Comment (nabobalis) [reply sunkit_image/radial.py]: No, I was just confused.

File: sunkit_image/radial.py [hunk 22]
```diff
@@ -10,6 +10,7 @@
 
 import sunpy.map
 from sunpy.coordinates import frames
+from sunpy.map import Map
```
Comment (nabobalis) [inline sunkit_image/radial.py:13]:
```suggestion
```
This has been imported above.

**Comment by nabobalis** [comment]
Thanks for the PR @GillySpace27, I noticed some effects due to upstream changes that I had to correct. 

I also added  some more changelogs due to the changes and some minor test fixes.

**Pull request merged**